dbx/sync/clients.py

```diff
@@ -0,0 +1,198 @@
+
+
+async def _api_delete(
+    *, api_base_path: str, path: str, session: aiohttp.ClientSession, recursive: bool = False, api_token: str
+):
+    dbx_echo(f"Deleting {path}")
+    json_data = dict(path=path)
+    if recursive:
+        json_data["recursive"] = recursive
+    while True:
+        headers = get_auth_headers(api_token)
+        async with session.post(url=f"{api_base_path}/delete", json=json_data, headers=headers) as resp:
+            if resp.status in {200, 404}:
+                break
+            if resp.status == 429:
+                dbx_echo("Rate limited")
+                await _rate_limit_sleep(resp)
+            else:
+                txt = await resp.text()
+                dbx_echo(f"HTTP {resp.status}: {txt}")
+                raise ClientError(resp.status)
+
+
+async def _api_mkdirs(*, api_base_path: str, path: str, session: aiohttp.ClientSession, api_token: str):
+    dbx_echo(f"Creating {path}")
+    json_data = dict(path=path)
+    while True:
+        headers = get_auth_headers(api_token)
+        async with session.post(url=f"{api_base_path}/mkdirs", json=json_data, headers=headers) as resp:
+            if resp.status == 200:
+                break
+            if resp.status == 429:
+                dbx_echo("Rate limited")
+                await _rate_limit_sleep(resp)
+            else:
+                txt = await resp.text()
+                dbx_echo(f"HTTP {resp.status}: {txt}")
+                raise ClientError(resp.status)
+
+
+def check_path(path: str) -> None:
+    if not path:
+        raise ValueError("Path is empty")
+    if "\\" in path:
+        raise ValueError("Paths should not contain backslashes")
+
+
+class DBFSClient(BaseClient):
+    name = "dbfs"
+
+    def __init__(self, *, base_path: str, config: DatabricksConfig):
+        check_path(base_path)
+        self.base_path = "dbfs:" + base_path.rstrip("/")
+        self.api_token = config.token
+        self.host = config.host.rstrip("/")
+        self.api_base_path = f"{self.host}/api/2.0/dbfs"
+
+        dbx_echo(f"Target base path: {self.base_path}")
+
+    async def delete(self, sub_path: str, *, session: aiohttp.ClientSession, recursive: bool = False):
+        check_path(sub_path)
+        path = f"{self.base_path}/{sub_path}"
+        await _api_delete(
+            api_base_path=self.api_base_path, path=path, session=session, recursive=recursive, api_token=self.api_token
+        )
+
+    async def mkdirs(self, sub_path: str, *, session: aiohttp.ClientSession):
+        check_path(sub_path)
+        path = f"{self.base_path}/{sub_path}"
+        await _api_mkdirs(api_base_path=self.api_base_path, path=path, session=session, api_token=self.api_token)
+
+    async def put(
+        self, sub_path: str, full_source_path: str, *, session: aiohttp.ClientSession, overwrite: bool = True
+    ):
+        check_path(sub_path)
+        path = f"{self.base_path}/{sub_path}"
+        dbx_echo(f"Putting {path}")
```

Comment:
Yes, originally I used `logging`, however when integrating this code with dbx I chose to switch everything over to `dbx_echo` because the logging was not showing up (only ERRORS are logged).  This is something I'd like to address in a later PR.  I didn't want to change how logging is configured yet.